File: 75 must do questions/125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#

# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s: return True
        left, right = 0, len(s)-1
        
        while left < right:
            if s[left].isalnum() and s[right].isalnum():
                if s[left].lower() != s[right].lower():
                    return False
                left +=1; right -= 1

            else:
                if not s[left].isalnum():
                    left+=1
                if not s[right].isalnum():
                    right-=1

        return True

        
# @lc code=end

File: 125.valid-palindrome.java
/*
 * @lc app=leetcode id=125 lang=java
 *
 * [125] Valid Palindrome
 */

// @lc code=start
class Solution {

  public boolean isPalindrome(String s) {
    if (s.isEmpty()) {
      return true;
    }

    // Normalize string
    String normalized = s
      .replaceAll("[^a-zA-Z0-9]", " ")
      .replaceAll("\\s+", "")
      .toLowerCase();
    System.out.println(normalized);

    // Check if it's a palindrome
    for (int i = 0, j = normalized.length() - 1; i < j; i++, j--) {
      System.out.println(
        "i = " + i + normalized.charAt(i) + "\nj = " + j + normalized.charAt(j)
      );
      if (normalized.charAt(i) != normalized.charAt(j)) {
        return false;
      }
    }
    return true;
  }
}
// @lc code=end
